### PR TITLE
Use doctrine in the installer

### PR DESCRIPTION
**Pull request by CHItA**
Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16643

### CODE REVIEW

File: phpBB/composer.json [hunk 1]
```diff
@@ -25,14 +25,15 @@
 		"phpbb/phpbb-core": "self.version"
 	},
 	"require": {
-		"php": "^7.1.3",
+		"php": ">=7.3.0",
```
Comment (marc1706) [inline phpBB/composer.json:28]: How about `^7.3 || ^8.0` for now?

File: phpBB/phpbb/db/doctrine/connection_factory.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+	/**
+	 * Creates a Doctrine DBAL connection from phpBB configuration.
+	 *
+	 * @param config_php_file $config Config PHP file wrapper.
+	 *
+	 * @return Connection Doctrine DBAL connection.
+	 *
+	 * @throws runtime_exception		If the database connection could not be established.
+	 * @throws InvalidArgumentException	If $driver_name is not a valid phpBB database driver.
```
Comment (marc1706) [inline phpBB/phpbb/db/doctrine/connection_factory.php:38]: Did you mean `$driver`?

File: phpBB/phpbb/db/doctrine/connection_factory.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	 * @param string		$host		Hostname.
+	 * @param string|null	$user		Username.
+	 * @param string|null	$password	Password.
+	 * @param string|null	$name		Database name.
+	 * @param string|null	$port		Database port.
+	 *
+	 * @return Connection Doctrine DBAL connection.
+	 *
+	 * @throws runtime_exception		If the database connection could not be established.
+	 * @throws InvalidArgumentException	If $driver_name is not a valid phpBB database driver.
```
Comment (marc1706) [inline phpBB/phpbb/db/doctrine/connection_factory.php:72]: Same here

File: phpBB/phpbb/db/doctrine/connection_parameter_factory.php [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+				'charset' => 'UTF8',
+			],
+			'pdo_pgsql' => [
+				'charset' => 'UTF8',
+			],
+		];
+
+		if ($params['driver'] === 'pdo_mysql')
+		{
+			$enrichment_tags['pdo_mysql'][\PDO::MYSQL_ATTR_FOUND_ROWS] = true;
```
Comment (marc1706) [inline phpBB/phpbb/db/doctrine/connection_parameter_factory.php:162]: I think we should also add `"ext-pdo": "*"` to composer.json

File: phpBB/phpbb/db/doctrine/driver_convertor.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+		$name = str_replace('phpbb\db\driver', '', $driver_name);
+		$name = preg_replace('/mysql$/i', 'mysqli', $name);
+		$name = trim($name, '\\');
+
+		switch ($name)
+		{
+			case 'mssql_odbc':
+			case 'mssqlnative':
+				$name = 'pdo_sqlsrv';
+				break;
```
Comment (marc1706) [inline phpBB/phpbb/db/doctrine/driver_convertor.php:44]: `break;` should be aligned with `case`

File: phpBB/phpbb/install/database_task.php [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+use phpbb\install\helper\config;
+use phpbb\install\helper\database;
+use phpbb\install\helper\iohandler\iohandler_interface;
+
+/**
+ * Abstract base class for common database manipulation tasks.
+ */
+abstract class database_task extends task_base
+{
+	/**
```
Comment (marc1706) [inline phpBB/phpbb/install/database_task.php:32]: I think we can reduce these to e.g. `/** @var Connection */`

File: phpBB/phpbb/install/module/install_database/task/add_default_data.php [hunk 7]
```diff
@@ -53,37 +56,32 @@ class add_default_data extends \phpbb\install\task_base
...
-								\phpbb\language\language $language,
-								$root_path)
+	public function __construct(database $db_helper,
+								config $config,
+								iohandler_interface $iohandler,
+								language $language,
+								string $root_path)
 	{
-		$this->db				= $container->get('dbal.conn.driver');
+		$this->db 				= self::get_doctrine_connection($db_helper, $config);
```
Comment (marc1706) [inline phpBB/phpbb/install/module/install_database/task/add_default_data.php:71]: I've spotted a space :)

**Comment by marc1706** [comment]
Apart from those and maybe some manual testing it looks pretty good.

File: phpBB/includes/questionnaire/questionnaire.php [hunk 8]
```diff
@@ -258,7 +258,7 @@ function get_data()
 		extract($phpbb_config_php_file->get_all());
 		unset($dbhost, $dbport, $dbname, $dbuser, $dbpasswd); // Just a precaution
 
-		$dbms = $phpbb_config_php_file->convert_30_dbms_to_31($dbms);
+		$dbms = \phpbb\config_php_file::convert_30_dbms_to_31($dbms);
```
[the remark this answers is not in the document]
Comment (CHItA) [reply phpBB/includes/questionnaire/questionnaire.php:261]: Yes and no. You need this if you don't change the driver name in `config.php`, which would need a migration. Note, that after something like #5841 you don't need the 3.1+ driver name neither, at which point I would suggest adding a migration to record the doctrine driver name instead. Two migrations seem a bit much for this IMO, and these are not that hard to find and remove later.

Removing our own DBAL is not all that much work based on the PR mentioned before, and IIRC @mrgoldy has a branch for that as well. I would remove this in that PR.

**Review by marc1706 — APPROVED** [review]

**Pull request merged**